src/Distribution/Server/Features/Vouch.hs

```diff
@@ -0,0 +1,218 @@
+import Distribution.Server.Framework (abstractAcidStateComponent, emptyHackageFeature, errBadRequest)
+import Distribution.Server.Framework (featureDesc, featureReloadFiles, featureResources, featureState)
+import Distribution.Server.Framework (liftIO, makeAcidic, openLocalStateFrom, query, queryState, resourceAt, resourceDesc, resourceGet)
+import Distribution.Server.Framework (resourcePost, toResponse, update, updateState)
+import Distribution.Server.Framework.BackupRestore (RestoreBackup(..))
+import Distribution.Server.Framework.Templating (($=), TemplateAttr, getTemplate, loadTemplates, reloadTemplates, templateUnescaped)
+import qualified Distribution.Server.Users.Group as Group
+import Distribution.Server.Users.Types (UserId(..), UserInfo, UserName(..), userName)
+import Distribution.Server.Features.Upload(UploadFeature(..))
+import Distribution.Server.Features.Users (UserFeature(..))
+import Distribution.Simple.Utils (toUTF8LBS)
+
+newtype VouchData = VouchData (Map.Map UserId [(UserId, UTCTime)])
+  deriving (Show, Eq)
+  deriving newtype MemSize
+
+putVouch :: UserId -> (UserId, UTCTime) -> Update VouchData ()
+putVouch vouchee (voucher, now) = do
+  VouchData tbl <- get
+  let oldMap = fromMaybe [] (Map.lookup vouchee tbl)
+      newMap = (voucher, now) : oldMap
+  put $ VouchData (Map.insert vouchee newMap tbl)
+
+getVouchesFor :: UserId -> Query VouchData [(UserId, UTCTime)]
+getVouchesFor needle = do
+  VouchData tbl <- ask
+  pure . fromMaybe [] $ Map.lookup needle tbl
+
+getVouchesData :: Query VouchData VouchData
+getVouchesData = ask
+
+replaceVouchesData :: VouchData -> Update VouchData ()
+replaceVouchesData = put
+
+$(deriveSafeCopy 0 'base ''VouchData)
+
+makeAcidic ''VouchData
+  [ 'putVouch
+  , 'getVouchesFor
+  -- Stock
+  , 'getVouchesData
+  , 'replaceVouchesData
+  ]
+
+vouchStateComponent :: FilePath -> IO (StateComponent AcidState VouchData)
+vouchStateComponent stateDir = do
+  st <- openLocalStateFrom (stateDir </> "db" </> "Vouch") (VouchData mempty)
+  let initialVouchData = VouchData mempty
+      restore =
+        RestoreBackup
+          { restoreEntry = error "Unexpected backup entry"
+          , restoreFinalize = return initialVouchData
+          }
+  pure StateComponent
+    { stateDesc = "Keeps track of vouches"
+    , stateHandle = st
+    , getState = query st GetVouchesData
+    , putState = update st . ReplaceVouchesData
+    , backupState = \_ _ -> []
+    , restoreState = restore
+    , resetState = vouchStateComponent
+    }
+
+data VouchFeature =
+  VouchFeature
+    { vouchFeatureInterface :: HackageFeature
+    }
+
+instance IsHackageFeature VouchFeature where
+  getFeatureInterface = vouchFeatureInterface
+
+requiredCountOfVouches :: Int
+requiredCountOfVouches = 3
+
+isWithinLastMonth :: UTCTime -> (UserId, UTCTime) -> Bool
+isWithinLastMonth now (_, vouchTime) =
+  addUTCTime (30 * nominalDay) vouchTime < now
+
+data Err
```

Comment:
I'd write this out as a full word, eg VouchError, but I think the entire data structure and judge function is superflous.